### PR TITLE
internal/exec: don't relabel a mountpoint that already exists

### PR DESCRIPTION
**Pull request by Adam0Brien**
Fix: #1452

### CODE REVIEW

**Comment by Adam0Brien** [comment]
Marking as ready for review as all tests are passing.

File: docs/release-notes.md [hunk 1]
```diff
@@ -19,6 +19,7 @@ nav_order: 9
 
 - Document that `hash` fields describe decompressed data
 - Correctly document Tang `advertisement` field as optional
+- Fix relabling mountpoints which already exist
```
Comment (prestist) [inline docs/release-notes.md:22]: hmm that verbiage sounds like we are allowing the relabeling of mountpoints which already exist. 

WDYT about
"Correctly avoid relabeling a mount point that exists"

File: internal/exec/stages/mount/mount.go [hunk 2]
```diff
@@ -142,31 +142,36 @@ func (s stage) mountFs(fs types.Filesystem) error {
...
 	}
 
-	if distro.SelinuxRelabel() {
-		if err := s.RelabelFiles([]string{firstMissing}); err != nil {
-			return err
+	isReadOnly := util.FileSystemIsReadOnly(path)
+	if isReadOnly {
+		fmt.Println("File system is read-only")
+	} else {
+
```
Comment (prestist) [inline internal/exec/stages/mount/mount.go:149]: We have to save space... Lol jk but I dont think we need this blank space.

File: internal/exec/stages/mount/mount.go [hunk 3]
```diff
@@ -142,31 +142,36 @@ func (s stage) mountFs(fs types.Filesystem) error {
 		return err
 	}
 
-	if distro.SelinuxRelabel() {
-		if err := s.RelabelFiles([]string{firstMissing}); err != nil {
-			return err
+	isReadOnly := util.FileSystemIsReadOnly(path)
+	if isReadOnly {
+		fmt.Println("File system is read-only")
```
Comment (prestist) [inline internal/exec/stages/mount/mount.go:147]: hmm, why are we if/elseing here and what do we want to do with the `println`?

File: tests/positive/filesystems/skip_readonly.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+import (
+	"github.com/coreos/ignition/v2/tests/register"
+	"github.com/coreos/ignition/v2/tests/types"
+)
+
+func init() {
+	register.Register(register.PositiveTest, SkipReadOnlyFilesystems())
+}
+
+func SkipReadOnlyFilesystems() types.Test {
```
Comment (prestist) [inline tests/positive/filesystems/skip_readonly.go:26]: how about a negative test as well ?

**Review by prestist — COMMENTED** [review]
Great first attempt!

A few changes, and I noticed that your commit message was a bit long, and might need a few more details

**Comment by prestist** [comment]
You will also need to rebase off of main to get your CI working due to https://github.com/coreos/ignition/pull/1621

File: tests/positive/filesystems/skip_readonly.go (hunk 4 above)
Comment (Adam0Brien) [reply tests/positive/filesystems/skip_readonly.go:26]: sounds good ill take a look at making one!

File: docs/release-notes.md (hunk 1 above)
Comment (Adam0Brien) [reply docs/release-notes.md:22]: SGTM!

**Comment by prestist** [comment]
Could we squash all these commits into one please?

File: internal/exec/stages/mount/mount.go [hunk 5]
```diff
@@ -142,31 +142,35 @@ func (s stage) mountFs(fs types.Filesystem) error {
 		return err
 	}
 
-	if distro.SelinuxRelabel() {
-		if err := s.RelabelFiles([]string{firstMissing}); err != nil {
-			return err
+	isReadOnly := util.FileSystemIsReadOnly(path)
```
Comment (bgilbert) [inline internal/exec/stages/mount/mount.go:145]: This seems like a pretty indirect way to go about things.  If the filesystem is read-only and we needed to create the mountpoint directory, we already failed to do that before we got here.  If the mountpoint directory already existed, we'll relabel the mountpoint _most of the time_, but silently skip doing so if the filesystem is read-only.  That special case feels wrong - it could surprise a user who's expecting their mountpoint to be relabeled - so it's worth stepping back and looking for a different way to solve this.

The problem as described in https://github.com/coreos/ignition/issues/1452#issuecomment-1224386412 is that `FindFirstMissingPathComponent` doesn't distinguish whether the specified path already exists or not.  If the directory already exists, maybe we don't need to relabel it at all?

[Blame](https://github.com/coreos/ignition/blame/24c18840349aa9b981055e12f2ec028b3e8a609f/internal/exec/stages/mount/mount.go#L145) points to 381132ff15f30744b6af2a05fea85e3d0ca1cb5b, whose commit message says it wants to relabel `the mount point we create`.  And the code from that commit calls a function named `FindFirstMissingDirForFile`, which is consistent with the idea that it wants to relabel a _newly-created_ directory.  But that's not what the function is named now.  Looking at [blame](https://github.com/coreos/ignition/blame/24c18840349aa9b981055e12f2ec028b3e8a609f/internal/exec/stages/mount/mount.go#L129) again, we find the function rename in e30dfa4add1b0ffe8aef7ce0ddd1721489ba07f6.  And sure enough, that commit clarifies the function's behavior if all directory components exist.  So it seems like the original commit was based on an incomplete analysis of the "already exists" case.

On the basis of that record, I think the right approach is to relabel only if the directory is newly created.  Since creating the directory can't succeed on a read-only filesystem, that'll solve #1452.  I think @jlebon, the author of 381132ff15f30744b6af2a05fea85e3d0ca1cb5b, will agree with that analysis, since he retitled #1452 to `Don't try to relabel mountpoints which already exist` last August. :slightly_smiling_face:

We're already checking whether the mountpoint exists on line 135, so this should be a fairly small change to the original code.

File: docs/release-notes.md [hunk 6]
```diff
@@ -26,6 +26,7 @@ Starting with this release, ignition-validate binaries are signed with the
 - Clarify documentation of `passwordHash` fields
 - Correctly document Tang `advertisement` field as optional
 - Fix failure disabling nonexistent unit with systemd ≥ 252
+- Correctly avoid relabeling a mount point that exists
```
Comment (bgilbert) [inline docs/release-notes.md:29]: "Correctly" is a bit strong.  Relabeling is probably harmless most of the time.  Maybe:

```suggestion
- Don't relabel a mount point that already exists
```

File: tests/positive/filesystems/skip_readonly.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+				"format": "ext4",
+				"wipeFilesystem": true,
+				"label": "READ-ONLY"
+			  }]
+		}
+	}`
+
+	in[0].Partitions.GetPartition("READ-ONLY").FilesystemType = "ext4"
+	out[0].Partitions.GetPartition("READ-ONLY").FilesystemType = "ext4"
+	in[0].Partitions.GetPartition("READ-ONLY").FilesystemMode = 0444
```
Comment (bgilbert) [inline tests/positive/filesystems/skip_readonly.go:53]: This just removes the write bit from the directory permissions, but Ignition runs as the root user, and root ignores permission bits.  A read-only filesystem is one mounted with `-o ro`.

And in any event, we need the filesystem containing the mountpoint to be read-only, not the one being mounted.

File: tests/types/types.go [hunk 8]
```diff
@@ -194,6 +195,15 @@ func GetBaseDisk() []Disk {
 					Length:         262144,
 					FilesystemType: "ext4",
 				},
+				{
+					Number:         12,
+					Label:          "READ-ONLY",
```
Comment (bgilbert) [inline tests/types/types.go:200]: This adds a new partition to every single test, which isn't necessary, and seems like it might affect other tests.  Let's just add the affected partition (or, more likely, reconfigure an existing one) in the new tests.

File: tests/positive/filesystems/skip_readonly.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+		{
+			Label:        "OEM",
+			Substitution: "$DEVICE",
+		},
+	}
+	config := `{
+		"ignition": {"version": "$version"},
+		"storage": {
+			"filesystems": [{
+				"path": "/dev/sda1",
```
Comment (bgilbert) [inline tests/positive/filesystems/skip_readonly.go:42]: This is using `/dev/sda1` as the mountpoint, which could be dangerous if the tests run on a system where `/dev/sda1` is a real disk partition.  It's also not what we need here; the mountpoint should be an existing directory inside a read-only filesystem.

**Review by bgilbert — COMMENTED** [review]
Thanks for the PR.  The release note is essentially correct (one small nit), but unfortunately the rest of the PR does something different than what the release note claims to do.  And also, the PR doesn't appear to have been tested against the bug it's trying to fix.  Details in comments below.

Blackbox tests have some runtime cost, and are mainly used to validate core functionality.  This change is a small fix for a corner case; blackbox tests seem too expensive (and not trivial to set up in this case) and unit tests can't check this.  I'd be inclined to recommend not adding tests for this one, and just manually testing that it fixes the bug.

**Comment by bgilbert** [comment]
You'll need to rebase to pick up CI changes.

File: tests/types/types.go (hunk 8 above)
Comment (Adam0Brien) [reply tests/types/types.go:200]: Makes sense ill drop in the next push!
ill have a go at re configuring an existing disk.

File: tests/positive/filesystems/skip_readonly.go (hunk 9 above)
Comment (Adam0Brien) [reply tests/positive/filesystems/skip_readonly.go:42]: I've seen other tests in /positive/filesystems use ```/tmp0``` for the path, will keep this in mind the next time i might have to write a Blackbox test
for now ill drop the negative and positive test

File: internal/exec/stages/mount/mount.go (hunk 5 above)
Comment (Adam0Brien) [reply internal/exec/stages/mount/mount.go:145]: Okay, so correct me if i'm wrong but it seems like i just need to write a function to check if the specified directory exists and if it does exist do nothing?

File: internal/exec/util/selinux.go [hunk 10]
```diff
@@ -84,3 +84,13 @@ func (ut Util) RelabelFiles(patterns []string) error {
 	}
 	return nil
 }
+
+func FileSystemIsReadOnly(path string) bool {
+	//open file for writing
+	file, err := os.OpenFile(path, os.O_WRONLY, 0666)
+	if err != nil {
+		return os.IsPermission(err)
```
[the remark this answers is not in the document]
Comment (Adam0Brien) [reply internal/exec/util/selinux.go:92]: Wasn't exactly sure how to test when writing this at first, 

but after reading this i think this is what your looking for:
`func isReadOnlyFileSystem(err error) bool {
	if pe, ok := err.(*os.PathError); ok {
		if errno, ok := pe.Err.(syscall.Errno); ok {
			return errno == syscall.EROFS || errno == syscall.EISDIR
		}
	}
	return false
}`

This function will return true if the directory is attempted to be open as a file or if the fs is read only

File: tests/positive/filesystems/skip_readonly.go (hunk 7 above)
Comment (Adam0Brien) [reply tests/positive/filesystems/skip_readonly.go:53]: Okay didn't know ignition ran as a root user,
I thought i could just assign the permission bits to a variable and check against them with the partition i made in types.go

This will be dropped in the next push since Blackbox tests aren't needed

File: internal/exec/stages/mount/mount.go (hunk 5 above)
Comment (bgilbert) [reply internal/exec/stages/mount/mount.go:145]: As I said, we're already doing this on line 135, so we should just be able to move some code under that conditional.

File: internal/exec/util/selinux.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (bgilbert) [reply internal/exec/util/selinux.go:92]: That's true, it will.  But you're not getting EISDIR as a valid indication that the filesystem is read-only, you're getting it because your code is wrong.  It doesn't make sense to try to open a directory as though it were a file.

You could try to create a temporary file _inside_ the directory.  But that still has a bit of a code smell, because you'd be changing something solely to see whether you have permission to change it.  The better approach is to query the filesystem with [`statfs(2)`](https://man7.org/linux/man-pages/man2/statfs.2.html) and check for the `ST_READONLY` flag.  Unfortunately, while the `statfs` syscall is available in `x/sys`, the value of `ST_READONLY` doesn't seem to be available anywhere in the Go standard libraries, so we'd need a bit of C glue for this.

But luckily none of that is relevant here, since we're [not going to be doing this check](https://github.com/coreos/ignition/pull/1608#discussion_r1222362864).

File: internal/exec/stages/mount/mount.go (hunk 5 above)
Comment (Adam0Brien) [reply internal/exec/stages/mount/mount.go:145]: okay that makes sense, and how long should a directory be considered newly created?
should it be considered "old" if its existed longer than x amount of minutes?

Comment (bgilbert) [reply internal/exec/stages/mount/mount.go:145]: "Newly created" means we just created it.  If Ignition creates the directory, we should relabel it, and if the directory is already there, we shouldn't.

File: internal/exec/stages/mount/mount.go [hunk 11]
```diff
@@ -133,6 +135,9 @@ func (s stage) mountFs(fs types.Filesystem) error {
 	}
 
 	if _, err := os.Stat(path); err != nil && os.IsNotExist(err) {
+		if _, err := util.PathExists(path); err != nil {
```
Comment (bgilbert) [inline internal/exec/stages/mount/mount.go:138]: The previous line verifies that the path does _not_ exist, so this condition will never be true.  Did you test this?

File: internal/exec/stages/mount/mount.go [hunk 12]
```diff
@@ -133,6 +135,9 @@ func (s stage) mountFs(fs types.Filesystem) error {
...
+			pathExists = true
```
Comment (bgilbert) [inline internal/exec/stages/mount/mount.go:139]: We don't need to introduce a variable for this, and it makes the code harder to follow.  Just move the relabeling call directly after the directory creation in the "doesn't exist" path.

And actually, `firstMissing` is only used during the directory creation, so its calculation can be moved into the "doesn't exist" path as well.

File: internal/exec/stages/mount/mount.go [hunk 13]
```diff
@@ -41,6 +41,8 @@ const (
 	name = "mount"
 )
 
+var pathExists = false
```
Comment (bgilbert) [inline internal/exec/stages/mount/mount.go:44]: Since this is a global variable, if it's set once, it'll affect every subsequent `mountFs()` call, which is certainly not what we want.

File: internal/exec/stages/mount/mount.go [hunk 14]
```diff
@@ -166,7 +170,6 @@ func (s stage) mountFs(fs types.Filesystem) error {
 			}
 		}
 	}
-
```
Comment (bgilbert) [inline internal/exec/stages/mount/mount.go:169]: Let's avoid gratuitous whitespace changes please.

**Review by bgilbert — COMMENTED** [review]
Your commits don't appear to be independent changes, so please squash them.

File: internal/exec/stages/mount/mount.go (hunk 13 above)
Comment (Adam0Brien) [reply internal/exec/stages/mount/mount.go]: makes perfect sense didn't realise that at the time thanks!

File: internal/exec/stages/mount/mount.go (hunk 11 above)
Comment (Adam0Brien) [reply internal/exec/stages/mount/mount.go]: I thought since IsNotExist had err being passed to it it wasn't checking paths, thanks for clearing that up!

I wasn't too sure how to test it at the time but i think i get the idea now i ran ./test on the most recent push and everything passed

File: internal/exec/stages/mount/mount.go (hunk 12 above)
Comment (Adam0Brien) [reply internal/exec/stages/mount/mount.go]: makes sense have that changed now!

**Comment by bgilbert** [comment]
Code LGTM.  Please update the commit message to correctly describe the change (read-only filesystems aren't relevant here) and link to the issue.

**Review by bgilbert — APPROVED** [review]
:tada:
